server/svix-server/src/queue/redis.rs

```diff
@@ -166,41 +313,72 @@ fn from_redis_key(key: &str) -> TaskQueueDelivery {
 impl TaskQueueSend for RedisQueueProducer {
     async fn send(&self, task: QueueTask, delay: Option<Duration>) -> Result<()> {
         let mut pool = self.pool.get().await.unwrap();
+
+        // If there's a delay, compute the timestamp at which the delay is up
         let timestamp = delay.map(|delay| Utc::now() + chrono::Duration::from_std(delay).unwrap());
-        let delivery = TaskQueueDelivery::new(task, timestamp);
-        let key = to_redis_key(&delivery);
+
         if let Some(timestamp) = timestamp {
+            // If there's a delay, add it to the DELAYED queue by ZADDING the Redis-key-ified
+            // delivery
+            let delivery = TaskQueueDelivery::new(task, Some(timestamp));
+            let key = to_redis_key(&delivery);
             let _: () = pool
-                .zadd(DELAYED, key, timestamp.timestamp())
+                .zadd(self.delayed_queue_name, key, timestamp.timestamp())
                 .await
                 .unwrap();
         } else {
-            let _: () = pool.rpush(MAIN, key).await.unwrap();
+            // If there is no delay simply XADD the task to the MAIN queue
+            let _: () = pool
+                .query_async(Cmd::xadd(
+                    &self.main_queue_name,
+                    GENERATE_STREAM_ID,
+                    &[(QUEUE_KV_KEY, serde_json::to_string(&task).unwrap())],
+                ))
+                .await
+                .unwrap();
         }
         tracing::trace!("RedisQueue: event sent > (delay: {:?})", delay);
         Ok(())
     }
 
     async fn ack(&self, delivery: TaskQueueDelivery) -> Result<()> {
-        let key = to_redis_key(&delivery);
+        // ACKing the delivery, XACKs the message in the queue so it will no longer be retried
         let mut pool = self.pool.get().await.unwrap();
-        let processed: u8 = pool.lrem(PROCESSING, 1, &key).await.unwrap();
+        let processed: u8 = pool
+            .query_async(Cmd::xack(
+                self.main_queue_name,
+                WORKERS_GROUP,
+                &[delivery.id.as_str()],
+            ))
+            .await
+            .unwrap();
         if processed != 1 {
             tracing::warn!(
-                "Expected to remove 1 from the list, removed {} for {}",
+                "Expected to remove 1 from the list, removed {} for {}|{}",
                 processed,
-                key
+                delivery.id,
+                serde_json::to_string(&delivery.task).unwrap()
             );
         }
         Ok(())
     }
 
     async fn nack(&self, delivery: TaskQueueDelivery) -> Result<()> {
-        // FIXME: do something else here?
-        let key = to_redis_key(&delivery);
+        // FIXME: NACKing will also XACK the message, but something else should be done here
```

Comment:
TBH, I think nacking should probably not do anything for now (other than print an error) as we probably just want it to be picked up again once the pending timeout finishes?